### ISSUE TITLE
Comparison of desktop app construction toolkits

### ISSUE THREAD

**Issue by adulbrich** [radiantlab/HDRICalibrationTool#8]
A few toolkits/frameworks to consider in this comparison:
- Tauri
- Electron
- Neutralinojs
- Photnio
- Revery

What to consider:
- Performance
- Adoption
- Support
- Ease-of-use
- Documentation
- Ease of packaging executables on all platforms
- ...

**Comment by shantimorrell** [comment]
I started on some research (plenty to go...)

I added a Google Doc with my current notes to our shared drive. Still very much a work in progress, but wanted to share it so you all can see it whenever you'd like.

[Link to notes](https://docs.google.com/document/d/1qcjqBR5iALWSiQhP61FbxWQ-qzsoSd6R7S7BwUzjwl8/edit)

**Comment by shantimorrell** [comment]
## Summary

Tauri and Electron appear to be the most viable options, with Neutralino also being a possibility. 

### Electron

Electron is definitely the most widely used and well-known desktop app toolkit, which is a significant advantage for our needs given most of our engineering team has limited experience with desktop development. Using a popular, well-established framework like Electron offers a much lower chance of encountering unforeseen bugs and significantly more detailed documentation than some of the most recent frameworks that are still in development. Besides the official documentation, there would also likely be the most numerous community posts regarding Electron compared to the newer, less-adopted frameworks. Additionally, we can be almost certain that Electron would offer whatever functionality we require. The other major benefit of Electron is that it uses HTML/CSS/JavaScript which our engineering team is at least somewhat familiar with.

The major drawback with Electron that might drive us to choose a different desktop app toolkit is the extensive size of the distributions and its memory usage during runtime. Electron includes a version of Node.js and a browser (Chromium) with the release of the app. Given that our app will likely be relatively small and simple, this may be something we’d like to avoid if possible. Electron is best for complex, heavy-weight applications, which does not align with our use case.

### Tauri

Tauri is the other most likely candidate. Tauri is known for creating light-weight, fast apps with relatively strong security. The smaller size of Tauri apps is due to shipping only the app without unnecessary binaries (Tauri does not include a browser). The small app size and strong security are also a result of using Rust for the backend, which is the feature that most distinguishes Tauri from other desktop app toolkits. Compared to JavaScript, Rust is more memory-efficient and secure. Additionally, Rust also offers performance benefits, particularly in terms of calls to the operating system and other backend functionality. Therefore, using Rust for the portions of our app that deal with file storage and handling the HDRGen and Radiance pipeline could offer a significant benefit to our app’s performance. Tauri allows Rust functions and programs to be called from the frontend.

However, the fact that Tauri uses Rust for its backend is also a potential drawback for our particular case because our engineering team does not have familiarity with Rust, and we would have to learn the language as we go. The other major concern with Tauri is its relative youth. While there’s been relatively high interest in Tauri in the last few years, the level of adoption by developers is still quite low compared to Electron. This could pose difficulties both in terms of finding documentation and support and in terms of discovering unforeseen bugs or lacking functionality. However, Tauri is one of the larger alternatives to Electron, so the resources available still appear fairly thorough.

### Neutralino

Neutralino is relatively similar to Electron in that it uses HTML/CSS/JavaScript, but in contrast offers lighter-weight, faster apps by using native browsers on each platform. The other major benefits of Neutralino are the relatively easy installation (no additional dependencies) and the ability to build the application for all platforms at once without using GitHub actions or installing additional packages.

Neutralino is still not widely adopted though and again this poses issues surrounding finding documentation and community posts and ensuring there’s adequate functionality for developing our app. The Neutralino docs do appear more detailed than some of the other lesser-known frameworks like Photino and Revery, but Neutralino’s documentation and community, and Neutralino as a whole, are still very immature compared to Electron.

### Other Alternatives

Photino and Revery are two other new frameworks that do offer the benefit of lighter-weight apps. However, they don’t appear to offer any significant advantage compared to the features offered in Electron, Tauri, or Neutralino, and their documentation is severely lacking. Both Photino and Revery are not well-established and are still in development. Given our team’s inexperience in developing desktop applications, there don’t appear to be enough benefits to justify using Photino or Revery over one of the other three choices: Tauri, Electron, or Neutralino.

### Overview

To summarize, Electron is likely the safest, most established option with the greatest documentation and community support, but it comes with less than ideal performance. 

Tauri is a newer alternative that still provides relatively decent documentation and offers a significant performance advantage over Electron, but would likely require learning some Rust for the backend.

Neutralino is probably the least safe option of these three frameworks given its immaturity and low adoption rate among developers, and is somewhat lacking in documentation, but does offer lighter-weight apps, an easier building process, and does not require learning Rust.

<br/>

![Image](https://github.com/radiantlab/HDRICalibrationTool/assets/91344787/51e80607-a01d-4cd3-bf3c-96c1502e0c30)

**Comment by adulbrich** [comment]
After discussion with the team, we're going with Tauri. Preference for writing the pipeline in Rust. 

Small proof-of-concept with Tauri/Svelte/TailwindCSS available here: https://github.com/adulbrich/hdr-image-viewer
